### ISSUE TITLE
Source Salesforce: Enable AvailabilityStrategy

### ISSUE THREAD

**Issue by lazebnyi** [airbytehq/airbyte#21785]

**Comment by grubberr** [comment]
I don't see reasons to enable AvailabilityStrategy for this connector and propose to keep it `None`.
**Source-Salesforce** already has all the needed logic to handle 403 and streams availability.

It's implemented using an optimistic strategy.
We try to synchronize streams and if one stream fails with 403 
we `raise AirbyteStopSync` exception and stop syncing the rest of the streams.

https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py#L144

Because the current implementation of AvailabilityStrategy works separately for every stream it even would not be possible to implement something similar using the current HttpAvailabilityStrategy.

That's the way I propose to keep the old implementation and don't use AvailabilityStrategy for this connector.

@lazebnyi @erohmensing Any comments?

**Comment by lazebnyi** [comment]
@grubberr  We check stream availability before syncing starts, and it's reasonable that we won't start syncing with problematic threads. But if we get any errors during synchronization, we should handle them, so in my opinion we should leave the current error handling and add enable AvailabilityStrategy.

@erohmensing WDYT?

**Comment by erohmensing** [comment]
To be clear, is 403 actually rate limiting when it comes to salesforce, and not a permissions error? If so, it probably doesn't belong on our permission failures dashboard, and yes we'd want to either not use or override the default behavior since the 403 meaning permission issues don't apply. 

For this line: 
```
if error.response.status_code == codes.FORBIDDEN and error_code == "REQUEST_LIMIT_EXCEEDED":
```
what if `error.response.status_code == codes.FORBIDDEN` but not `error_code == "REQUEST_LIMIT_EXCEEDED"`? Could that still mean a permissions error? If so, I think I'm with @lazebnyi - availabilitystrategy would check for the permissions error (we could customize it to check that the `error_code` is the permissions-related one, not one for rate limiting), and the current handling would deal with rate limiting

**Comment by erohmensing** [comment]
If we move forward with an implementation, might be worth discussing solutions for 
> it works but it slightly slows down "bulk" streams because bulk stream run an async job and wait the result

**Comment by lazebnyi** [comment]
As I understand after our last [conversation](https://airbytehq.slack.com/archives/C02U9R3AF37/p1676051790663069) in slack we can leave custom handling in read streams method. Other should be use AvailabilityStrategfy logic, right?

**Comment by erohmensing** [comment]
Yep, I think we want both - availabilitystrategy for stream-level permissions errors and the custom handling for the rate limiting max that could occur at any time during read. That's assuming they still do use 403s for permissions errors - if they use something else, we should customize to except the correct error code for permissions

**Comment by erohmensing** [comment]
> The source-salesforce connector already implements all logic that default HttpGeneralAvailability provides for us!

Indeed! which is great when we don't have wider handling for it - but now that we have something that can be reused across conntectors, we want to use that instead of customized code as much as possible. This lets us slowly move toward connectors being made up of mix-and-match components (think low-code) instead of custom code, which lets us change behavior and provide improvements across a wider number of connectors more easily. 

> But anyway it seems we can turn on HttpGeneralAvailability for all streams for case (1) only.
(For case (2) let the stream continue to work to stop the whole connector)

Sounds good to me, and let's remove the custom handling for case 1

> The downside of HttpGeneralAvailability is it will slow down the start of all streams, especially "bulk" streams.
I saw customers with about 900 streams.

Do you know how much it slows down the start of non-bulk streams? 

I don't think that the default makes sense for bulk streams, as it's essentially trying to get _all_ the data to check if its available before then grabbing the full data again. I'd keep it off for bulk data streams, maybe we can come up with something more generalized for that case.
